### PR TITLE
RUM-2528 fix: Remove weak UIViewController references

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Referencing and deallocating a `UIViewController` on a background thread can lead to the following exception:
```
"objc[55547]: Cannot form weak reference to instance (0x13f7fdc90) of class RedactedViewController. It is possible that this object was over-released, or is in the process of deallocation."
```

`UIViewController` deallocation doesn't seem to be thread-safe and so we should avoid referencing such instances.

### How?

Remove the `RUMViewIdentity` and only rely on hash for identifying a view, either when using a `UIViewController` or a key as `String`.

> We can no longer verify that a view controller is still allocated when resuming or creating upon expiration a session.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests for Core, RUM, Trace, Logs, CR and WVT
- [ ] Run unit tests for Session Replay
- [x] Run integration tests
- [ ] Run smoke tests
- [ ] Run tests for `tools/`

### CODE REVIEW

**Review by ncreated — COMMENTED** [review]
//

**Review by ncreated — CHANGES_REQUESTED** [review]
My biggest concern in this proposal is that we base identity on a wrong axiom. Quoting the `NSObject.hashValue` doc:
```
/// **Axiom:** `x == y` implies `x.hashValue == y.hashValue`
```
This implication isn't reverse, but we use it like it was 👎. In other words, in very rare situations the `hashValue` of two different identities might be the same.

What is worse, we base our solution on comparing `hashValue` on two different RUM view identities: `UIViewController.hashValue` and `String.hashValue`. What makes us thing that a certain hash of `UIViewController` will not collide with a hash of particular `String` used to key another RUM view?

Also, by removing `RUMViewIdentity` type and replacing it with basic `Int` we erase the origin of view identity, which will make debugging way more difficult. Imagine we now have `view.identifier = 42` - how would we know if it describes a `String` key or `UIViewController` instance?

Next, we assume that `UIViewController.hashValue` can be used to differentiate one instance of VC from another. Following code proves this assumption wrong:
```swift
class FooVC: UIViewController {
    override var hash: Int {  42 }
}
print(FooVC().hashValue) // 42
print(FooVC().hashValue) // 42
```

Last, by removing `RUMViewIdentity` type we lose opportunity of composing other identity-specific logics, like the construction of default `view.name`. Being isolated, it was very easy to unit test in `RUMViewIdentityTests`. By moving it to `Monitor`, we hidden this logic and lost tests coverage 👎.

The original problem can be solved within existing `RUMViewIdentity` implementation with making use of associated object and requiring minimal impact on the rest of the codebase:
```swift
extension UIViewController: RUMViewIdentifiable {
   func asRUMViewIdentity() -> RUMViewIdentity {
      if let existingIdentity = self.associatedRUMIdentity {
          return existingIdentity
      } else {
          let newIdentity = associateIdentity()
          return newIdentity
      }
   }
}
```

I'd prefer promoting solution that is idiomatic to existing concepts and requires minimal change instead of changing the core concepts. Let me know how it sounds @maxep 🙂

**Comment by maxep** [comment]
@ncreated , thanks for the review 🙏 

I don't necessarily agree with your points, the `RUMViewIdentity` is an unnecessary complication since we don't need to carry the weak ref, I prefer to clean it up. Also, without the reference, debugging and knowing from which vc is the identity from won't be possible.

> This implication isn't reverse, but we use it like it was 👎. In other words, in very rare situations the hashValue of two different identities might be the same.

What do you mean by that. If 2 hashes are different, we can consider the view as different 🤔 

> What makes us thing that a certain hash of UIViewController will not collide with a hash of particular String used to key another RUM view?

To me the answer is the range of `Int` (int-64), it's very unlikely that 2 objects will give the same hash, no? especially in a context of a session.

> we assume that `UIViewController.hashValue` can be used to differentiate one instance of VC from another. Following code proves this assumption wrong:

No, I would argue that overriding the `hash` is an explicit attempt on making instances identical so I wouldn't take that in consideration.

> by removing `RUMViewIdentity` type we lose opportunity of composing other identity-specific logics, like the construction of default `view.name`.

My reasoning here is more about the `name` and `path` being `nil` with default applied down the stream. It's preferable to fully construct the command at the higher level, making clear what value is applied to what property.

---

The PR is actually a second iteration of the solution, the first was still using a `enum ViewIdentifier` with associated `ObjectIdentifer` or `String`. I can push a commit to show you. This solution is more idiomatic IMO, better that using an objc associated object. Let me know if you prefer this!

**Review by ncreated — COMMENTED** [review]
>> This implication isn't reverse, but we use it like it was 👎. In other words, in very rare situations the hashValue of two different identities might be the same.

> What do you mean by that. If 2 hashes are different, we can consider the view as different 🤔

I mean the reverse situation. If 2 hashes are equal, we cannot consider views as the same. It is true that this situation is very unlikely, but it is also true that we based previous solution on wrong axiom.

> The PR is actually a second iteration of the solution, the first was still using a `enum ViewIdentifier` with associated `ObjectIdentifer` or `String`. I can push a commit to show you. This solution is more idiomatic IMO, better that using an objc associated object. Let me know if you prefer this!

What is proposed now, makes the proper use of `ObjectIdentifer` to distinguish one view identity from another. It also compares apples to apples, not apples to oranges like we proposed before (`String.hashValue` to `VC.hashValue`). The code is correct now 👍.

---

It looks good overall, but I left extra question on `view.path` (`.url`). It seems that we change it here with no reason 🤔💭.

File: DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift [hunk 1]
```diff
@@ -211,7 +203,7 @@ extension RUMViewsHandler: UIViewControllerHandler {
                 view: .init(
                     identity: identity,
                     name: rumView.name,
-                    path: rumView.path,
+                    path: rumView.path ?? viewController.canonicalClassName,
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift:206]: Probably a mistake from me, let me double check 👍

[the remark this answers is not in the document]
Comment (maxep) [reply DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift:206]: It's actually the same as before, in the [RUMStartViewCommand.init](https://github.com/DataDog/dd-sdk-ios/blob/develop/DatadogRUM/Sources/RUMMonitor/RUMCommand.swift#L76) the path default to the `canonicalClassName`, which is then used to fill the `view.url`. or I'm missing something?

File: DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift [hunk 2]
```diff
@@ -221,7 +213,7 @@ extension RUMViewsHandler: UIViewControllerHandler {
                 view: .init(
                     identity: identity,
                     name: "RUMUntrackedModal",
-                    path: nil,
+                    path: viewController.canonicalClassName,
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift:216]: See my response above. I don't intend to change anything in that regard, just the way the value are set due to the change of the `ViewIdentifier`. I also think it's clearer to see the path being set to the `UIViewController.canonicalClassName` explicitly. Let me know but I don't think I'm breaking the `view.url`.

**Review by ncreated — DISMISSED** [review]
LG 🙂👍

File: DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogRUM/Sources/Instrumentation/Views/RUMViewsHandler.swift:216]: Could be worthy having some unit tests to ensure no regression in the future. [Previously it was the part of `RUMViewIdentity` tests](https://github.com/DataDog/dd-sdk-ios/blob/5073cf69eb20163af30042a0a9e6e46790526f3b/DatadogRUM/Tests/RUMMonitor/Scopes/Utils/RUMViewIdentityTests.swift#L70-L76), removed in this PR. The new `ViewIdentifier` abstraction is not equal 1:1 with previous model (can be seen by inline `path:` changes) so we might be missing coverage.

**Review by ncreated — APPROVED** [review]

**Pull request merged**